google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java

```diff
@@ -1170,9 +1172,78 @@ public Boolean call() {
   public TableResult query(QueryJobConfiguration configuration, JobOption... options)
       throws InterruptedException, JobException {
     Job.checkNotDryRun(configuration, "query");
+
+    // If all parameters passed in configuration are supported by the query() method on the backend,
+    // put on fast path
+    QueryRequestInfo requestInfo = new QueryRequestInfo(configuration);
+    if (requestInfo.isFastQuerySupported()) {
+      String projectId = getOptions().getProjectId();
+      QueryRequest content = requestInfo.toPb();
+      return fastQuery(projectId, content, options);
+    }
+    // Otherwise, fall back to the existing create query job logic
     return create(JobInfo.of(configuration), options).getQueryResults();
   }
 
+  private TableResult fastQuery(
+      final String projectId, final QueryRequest content, JobOption... options)
+      throws InterruptedException {
+    com.google.api.services.bigquery.model.QueryResponse results;
+    try {
+      results =
+          runWithRetries(
+              new Callable<com.google.api.services.bigquery.model.QueryResponse>() {
+                @Override
+                public com.google.api.services.bigquery.model.QueryResponse call() {
+                  return bigQueryRpc.queryRpc(projectId, content);
+                }
+              },
+              getOptions().getRetrySettings(),
+              EXCEPTION_HANDLER,
+              getOptions().getClock());
+    } catch (RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+
+    // classic path
+    if (!results.getJobComplete() || results.getPageToken() != null) {
+      // Use jobId returned from backend to return full TableResult
+      String jobId = results.getJobReference().getJobId();
+      Job job = getJob(JobId.of(jobId));
+      TableResult result = job.getQueryResults();
+      return result;
+    } else { // fast path
```

Comment:
you don't need the else here since the last statement in the if branch is a "return result;"

You can thus eliminate that cide nesting and make the fast path

Reply:
done